### PR TITLE
M3-2045 Change: Refined Search POC

### PR DESCRIPTION
**Pull request by johnwcallahan**
## Description

This PR introduces new functionality to the search bar. Example queries:

`tag:my-tag`
`tag:my-tag OR tag:my-tag2`
`tag:my-tag,my-tag2` _(equivalent to `tag:my-tag AND tag:my-tag2`)_
`type:linode OR type:domain`
`-type:linode` (returns all types _except_ Linodes)

Nested queries are also possible:

`type:domain OR (type:linode AND tag:my-tag)`

A few aliases: 

- "&&" is equivalent to "AND"
- "||" is equivalent to "OR"
- "tags", "tag", and "group" will search the entity's tags.
- "label", "name", and "title" will search the entity's label

**Notes:** 
- Simple queries are still possible, of course. If no fields are specified, we search on label, tag, and ip address.

## Type of Change
- Non breaking change ('update', 'change')

## Note to Reviewers

The workings of this functionality are quite dense. I've added _heavy_ comments/documentation in `refinedSearch.ts`. Please let me know what can be clearer.

Please try to break the search bar best you can. 

**New libraries:**
- [logic-query-parser](https://www.npmjs.com/package/logic-query-parser)
- [search-string](https://www.npmjs.com/package/search-string)

### CODE REVIEW

File: src/features/Search/withStoreSearch.tsx [hunk 1]
```diff
@@ -31,27 +36,24 @@ export const search = (
     return entities; // could also return empty results, but this matches existing pattern.
   }
   const { linodes, volumes, domains, nodebalancers, images } = entities;
+
+  // Flatten all entities so we can search a single array. We add "entityType" to each item.
+  const entitiesToSearch = [
```
Comment (rjstires) [inline src/features/Search/withStoreSearch.tsx]: So everytime search is invoked, this array has to be created which involves mapping over all the things and updating them. Get the data into the appropriate state before it gets searched.

File: src/utilities/refinedSearch.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  return false;
+};
+
+// Takes a portion of a query ("tag:production") and an entity, and returns
+// TRUE if the query matches the entity. The query can match the entity by key type
+// (e.g. "tag:production,my-app" will look for entities with a tag called "production" AND a
+// tag called "my-app"). If there is no key specified, we'll search label, tags, and ipv4.
+const createCondition = (
```
Comment (martinmckenna) [inline src/utilities/refinedSearch.ts]: I'd like to see tests for every single function in the file, honestly. Looking at it from an outsider's perspective, it can be hard to follow and see the purpose of these functions and I definitely think tests would be a good reference point when coming back to this code months later.

**Comment by johnwcallahan** [comment]
> One tiny thing, which I don't think is much of a problem, is that for some reason `\` and `/` have different effects (one is a typo, the other seems to be a special case).

@Jskobos good catch; I hadn't noticed that. The library doing the operator-operand parsing is treating backslashes as a way to escape quotes. See: 

screen shot 2019-02-14 at 10 17 02 am

File: src/features/TopMenu/SearchBar/SearchBar.tsx [hunk 3]
```diff
@@ -147,6 +136,17 @@ class SearchBar extends React.Component<CombinedProps, State> {
         ? []
         : [defaultOption, ...combinedResults];
 
+    /* The final option in the list will be the "go to search results page" link.
+     * This doesn't share the same shape as the rest of the results, so should use
+     * the default styling. */
+    const Option = (props: any) => {
```
Comment (martinmckenna) [inline src/features/TopMenu/SearchBar/SearchBar.tsx]: what was the purpose of putting this inside the render method?

Comment (johnwcallahan) [reply src/features/TopMenu/SearchBar/SearchBar.tsx]: Oh, good catch! It was because `<SearchSuggestion />` needs the `searchText` prop. 

I just pushed up a fix, adding `searchResult` to each result (in the refinedSearch method, so we don't have to iterate over the results additional times).
